File: IT/Database/Esercizi/Java/Esercizio9.java
/*
file Esercizio9.java
eseguire con: java -cp .:sqlitejdbc-v<nnn>.jar Esercizio9
*/

import java.io.Console;
import java.sql.DriverManager;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.Statement;
import java.sql.PreparedStatement;
import java.sql.SQLException;

public class Esercizio9 {
	public static void main(String [] args) {
		System.out.println("Aggiorna il voto d'esame degli studenti.");

		Connection conn = null;
		Statement stmt = null;
		PreparedStatement pStmt = null;
		ResultSet res = null;
		int [] records = {};

		try {
			Class.forName("org.sqlite.JDBC");
			String dbName = "students.db";
			conn = DriverManager.getConnection("jdbc:sqlite:" + dbName);
			stmt = conn.createStatement();

			res = stmt.executeQuery("SELECT number, name, surname FROM students");

			if (!res.isBeforeFirst()) {
				System.out.println("Non ci sono studenti iscritti al corso!");
				res.close();
				stmt.close();
				conn.close();
				return;
			}

			pStmt = conn.prepareStatement("UPDATE students SET grade = ? WHERE number = ?");
			Console console = System.console();

			while (res.next()) {
				String num = res.getString(1);
				System.out.println("Matricola " + num + ", " + res.getString(2) + " " + res.getString(3));

				String s = console.readLine("Voto (da 18 a 30): ");
				if (s.isEmpty()) continue;
				int grade = 0;

				try {
					grade = Integer.parseInt(s);
					if (grade < 18 || grade > 30) {
						System.err.println("Voto non valido!");
						continue;
					}
				}
				catch (NumberFormatException e) {
					System.err.println(e);
					continue;
				}

				pStmt.setInt(1, grade);
				pStmt.setString(2, num);
				pStmt.addBatch();
				// pStmt.clearParameters();
			}

			records = pStmt.executeBatch();
			// pStmt.clearBatch();
		}
		catch (ClassNotFoundException e) {
			System.err.println(e);
		}
		catch (SQLException e) {
			System.err.println(e);
		}
		finally {
			try {
				if (res != null) res.close();
				if (stmt != null) stmt.close();
				if (pStmt != null) pStmt.close();
				if (conn != null) conn.close();
			}
			catch (SQLException e) {
				System.err.println(e);
			}
		}

		int tot = 0;
		for (int record : records) tot += record;

		System.out.println(tot + " studenti hanno passato la prova d'esame.");
	}
}

File: IT/Database/Esercizi/C++/esercizio9.cpp
/*
file esercizio9.cpp
compilare con: g++ -std=c++11 esercizio9.cpp
*/

#include <iostream>
#include <stdexcept>
#include <sqlite3.h>

using std::string;
using std::cin;
using std::cout;
using std::endl;
using std::invalid_argument;

int main()
{
	cout << "Aggiorna il voto d'esame degli studenti." << endl;

	sqlite3 *db;
	sqlite3_stmt *res;
	sqlite3_stmt *pStmt;
	sqlite3_open("students.db", &db);

	int rc = sqlite3_prepare_v2(db, "SELECT number, name, surname FROM students", -1, &res, NULL);

	if (rc != SQLITE_ERROR) {
		if (sqlite3_step(res) == SQLITE_DONE) {
			cout << "Non ci sono studenti iscritti al corso!" << endl;
			sqlite3_finalize(res);
			sqlite3_close(db);
			return 0;
		}

		sqlite3_prepare_v2(db, "UPDATE students SET grade = ? WHERE number = ?", -1, &pStmt, NULL);

		do {
			string num = (const char*) sqlite3_column_text(res, 0);
			string name = (const char*) sqlite3_column_text(res, 1);
			string surname = (const char*) sqlite3_column_text(res, 2);
			cout << "Matricola " << num << ", " << name << " " << surname << "." << endl;

			cout << "Voto (da 18 a 30): ";
			string s;
			getline(cin, s);

			if (s.empty()) continue;

			int grade = 0;

			try {
				grade = stoi(s);
				if (grade < 18 || grade > 30) {
					cout << "Voto non valido!" << endl;
					continue;
				}
			}
			catch (invalid_argument& e) {
				cout << e.what() << endl;
				continue;
			}

			if (grade != 0) {
				sqlite3_bind_int(pStmt, 1, grade);
			}
			else {
				sqlite3_bind_null(pStmt, 1);
			}

			sqlite3_bind_text(pStmt, 2, num.c_str(), -1, SQLITE_STATIC);

			if (sqlite3_step(pStmt) != SQLITE_DONE) {
				cout << sqlite3_errmsg(db) << endl;
			}

			sqlite3_reset(pStmt);
			sqlite3_clear_bindings(pStmt);

		} while (sqlite3_step(res) == SQLITE_ROW);

		int records = sqlite3_total_changes(db);
		cout << records << " studenti hanno passato la prova d'esame." << endl;
	}
	else {
		cout << sqlite3_errmsg(db) << endl;
	}

	sqlite3_finalize(res);
	sqlite3_finalize(pStmt);
	sqlite3_close(db);
}

